### ISSUE TITLE
Look into possibly supporting Stack

### ISSUE THREAD

**Issue by spinda** [ucsd-progsys/liquidhaskell#407]
Brief: https://www.reddit.com/r/haskell/comments/3a6p9h/how_stack_lowers_the_barrier_to_open_source/

[Stack](https://github.com/commercialhaskell/stack#readme) is a new build tool from the commercial-haskell project that might be worth looking into. It handles build caching, dependency management, multi-package projects (useful for our case), and so on, without the fuss of Cabal sandboxes, linking, and recompiles. Despite the name, it's not locked to Stackage, and can pull packages from both Hackage and arbitrary git commits.

**Comment by ranjitjhala** [comment]
Yes, I was also pondering this (having spent the better part of the last
few days
burning cycles recompiling sandboxes a billion times...)
​

**Comment by gridaphobe** [comment]
Hehe, I (mostly) haven't had to deal with cabal nonsense since I switched to using the nix package manager :)

Still, `stack` seems like a useful tool, and it looks like you just need to add a `stack.yaml` file with some configuration options. But I think if we're actually going to add a `stack.yaml` file to the repo, one of us should actually be using it. Otherwise it might bitrot..

**Comment by spinda** [comment]
I can try it out. I gave up on sandboxes and have just been cabal
install-ing everything, so stack would been an improvement.

**Comment by spinda** [comment]
I have a test of one method of setting up a workspace, in one repository with the other repositories as submodules:

https://github.com/spinda/liquid

After installing stack, running

```bash
git clone git@github.com:spinda/liquid.git
cd liquid
git submodule init
git submodule update
```

to set up the repository and

```bash
stack build
```

to build `liquid-fixpoint` and `liquidhaskell` worked perfectly. Then

```bash
stack exec -- liquid tests/pos/Abs.hs
stack exec -- liquid tests/neg/alias00.hs
```

to verify that running `liquid` worked.

`stack test` built everything successfully but failed in the test runner, which assumes the `liquid` binary will be located at `dist/build/liquid/liquid`.

Overall, a very nice first impression.

**Comment by spinda** [comment]
The "megarepository" model isn't required for using Stack, but putting a `stack.yaml` file in each repository doesn't work as nicely for development right now due to https://github.com/commercialhaskell/stack/issues/348.

**Comment by ranjitjhala** [comment]
In perfectly happy to use it, once we can set it up...

**Comment by gridaphobe** [comment]
@spinda where does stack put the `liquid` binary?

**Comment by spinda** [comment]
@gridaphobe `stack build` puts the `liquid` binary in `.stack-work/dist/x86_64-linux/Cabal-1.22.2.0/build/liquid/liquid` (on Linux x86_64 with Cabal 1.22.2). This is added to `$PATH` when you run `stack exec` or `stack test`. `stack install` puts it in `~/.local/bin/liquid`.

**Comment by gridaphobe** [comment]
Ugh, why can't they just leave a copy in `dist/build` where cabal puts it in the first place...

**Comment by ranjitjhala** [comment]
Regardless, why don't we just use find executable to get the path instead of using one of the other hardwired path ?

**Comment by gridaphobe** [comment]
Well then we have to worry about where `liquid` appears on the path. The testsuite wants the development binary, not some other version of `liquid` that you have installed.

**Comment by ranjitjhala** [comment]
Don't think I've ever had (or wanted) two versions :) 

Even if you did have a non dev version you could just nuke/move it when testing.

**Comment by gridaphobe** [comment]
Really?? I find it super useful to have the current release on my path because I know it works.

We could add `dist/build/liquid` to the path and then `findExecutable`, but honestly, this seems like a bug in `stack`..

**Comment by ranjitjhala** [comment]
Btw isn't this solved by using 

   "stack exec liquid"

Instead of one or other hardwired path to liquid?

**Comment by gridaphobe** [comment]
That presumes you have `stack` installed.

**Comment by ranjitjhala** [comment]
Yes, so we can just have command line flag, so:

  make test STACK=1

(does the above) and

  make test

is for your plain cabal thing. (Just like 'THREADS').

​

**Comment by spinda** [comment]
The repository-with-submodules setup is actually working pretty smoothly so far. It's useful to be able to track the state of combinations of repositories together, switching between "workspaces" all at once.

re: the test runner, I believe the following should always find the path of the `liquid` executable:

```haskell
import System.Environment
import System.FilePath

getLiquidPath :: IO FilePath
getLiquidPath = do
    testPath <- getExecutablePath
    return $ replaceDirectory (replaceFileName testPath "liquid") "liquid"
```

as the `test` executable is always located in `$builddir/test/test` with the `liquid` executable in `$builddir/liquid/liquid`.

**Comment by gridaphobe** [comment]
Ah, good call with `getExecutablePath`, that ought to do the trick.

**Comment by ranjitjhala** [comment]
​btw, whats the story with hdevtools / ghc-mod etc. and stack?​ Does it
know to look in the right package database? (Is there a package database
that it should look at?)

​

**Comment by christetreault** [comment]
I've been spending most of today trying to get this playing nice with emacs and ghc-mod, and it seems like it could be done. However, out of the box, it tries to use it's own built-in ghc, and expects you to do `stack ghc` to get at it. Unless you have ghc installed outside of stack, you won't be able to `ghc` from the command line.

There is a command `stack setup` that seems like it's supposed to put stack's ghc on the PATH, however it doesn't actually do it for me. It seems like stack keeps its own internal PATH, and while you can make it show it to you, it prints out a bunch of other stuff that you'd have to do some shell script monkeying to do anything useful with.

The bottom line is that without being able to call `ghc` from the command line, a lot of these tools don't work. You'd think you could just create an alias `stack ghc -> ghc`, but it seems that any command line arguments you pass to `stack ghc` go to stack and not ghc. You could manually put stack's ghc on the path, but the whole gimmick here is that it can swap this out at will depending on the project. I'd hate to have to detect this and fix it constantly.

I feel that they'll probably figure this stuff out, but for now I'm going to go back to banging my head against cabal.

**Comment by spinda** [comment]
@christetreault

I have GHC 7.10.1 installed through my OS's package manager. stack detects
it on my $PATH and uses it automatically. stack setup should only be
necessary if (a) you want to have stack manage your GHC installation for
you or (b) you want the project to be built with a different version of GHC
than the one installed globally.

What's the output of stack --version and cat ~/.stack/global/stack.yaml for
you? This might've been messed up for some versions prior to stack-0.1.

Also, regarding stack setup mode:

 - I believe it tells you what to add to your $PATH after stack setup
completes. If not, stack's GHC should be installed somewhere under
~/.stack/programs/.)
 - You need to separate arguments to GHC with -- when using stack ghc,
like stack
ghc -- X.hs -o x. If you have your $PATH set up, though, then stack ghc
shouldn't be necessary.

**Comment by spinda** [comment]
@christetreault

I have GHC 7.10.1 installed through my OS's package manager. `stack` detects
it on my `$PATH` and uses it automatically. `stack setup` should only be
necessary if (a) you want to have `stack` manage your GHC installation for
you or (b) you want the project to be built with a different version of GHC
than the one installed globally.

What's the output of `stack --version` and `cat ~/.stack/global/stack.yaml` for
you? This might've been messed up for some versions prior to `stack-0.1`.

Also, regarding `stack setup` mode:

 - I believe it tells you what to add to your `$PATH` after `stack setup`
completes. If not, `stack`'s GHC should be installed somewhere under
`~/.stack/programs/`.)
 - You need to separate arguments to GHC with `--` when using `stack ghc`,
eg. `stack
ghc -- X.hs -o x`. If you have your `$PATH` set up, though, then `stack ghc`
shouldn't be necessary.

**Comment by christetreault** [comment]
I purged all traces of ghc from my machine before installing stack. I then installed stack 0.1.0.0 fresh.

I understand that your OP is in regards to using stack alongside cabal and a pre-installed GHC. My post was my findings in using pure stack, with nothing else (letting it manage GHC). I saw in the documentation that stack can play nice with GHC, however I prefer to work with the system when trying a new tool. I don't really see much use in using stack, if you're not going to _use_ it. I might as well just use stackage with cabal and be done with it.

Regarding your last two points:

- This is my suspicion as well. The documentation is kind of sparse on this though. Everything I've read seems to imply that this should set my path for me (and TBH, I can't think of any reason why it wouldn't. Seems like the sort of thing I'd want to put in my `haskell-mode-hook`). I could set the path manually, but the path it gives is very version specific, and I'm concerned that I'm going to have to fix this every time it decides to upgrade it on me.

- This is good to know. However, haskell-mode and friends don't know this, and I don't have the time or energy to make them.

My impression is that this tool works well if you just use it from the command line. I'm sure editor integration could be done, and it'd probably be great. It might even work as is, but I'd be the one writing the tutorial on how to do it that everybody would be googling.

**Comment by ranjitjhala** [comment]
Btw, hdevtools support for stack is imminent:

https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/haskell-stack/Skfd3QGFzgk/obx7IwPsBqUJ

On Thu, Jun 25, 2015 at 5:49 PM, Michael Smith <notifications@github.com>
wrote:

> @christetreault
>
> I have GHC 7.10.1 installed through my OS's package manager. stack detects
> it on my $PATH and uses it automatically. stack setup should only be
> necessary if (a) you want to have stack manage your GHC installation for
> you or (b) you want the project to be built with a different version of GHC
> than the one installed globally.
>
> What's the output of stack --version and cat ~/.stack/global/stack.yaml for
> you? This might've been messed up for some versions prior to stack-0.1.
>
> Also, regarding stack setup mode:
>
> - I believe it tells you what to add to your $PATH after stack setup
> completes. If not, stack's GHC should be installed somewhere under
> ~/.stack/programs/.)
> - You need to separate arguments to GHC with -- when using stack ghc,
> like stack
> ghc -- X.hs -o x. If you have your $PATH set up, though, then stack ghc
> shouldn't be necessary.
>
> On Thu, Jun 25, 2015 at 2:21 PM, Chris Tetreault <notifications@github.com
> >
> wrote:
>
>
> > I've been spending most of today trying to get this playing nice with
> > emacs and ghc-mod, and it seems like it could be done. However, out of
> the
> > box, it tries to use it's own built-in ghc, and expects you to do stack
> > ghc to get at it. Unless you have ghc installed outside of stack, you
> > won't be able to ghc from the command line.
> >
> > There is a command stack setup that seems like it's supposed to put
> > stack's ghc on the PATH, however it doesn't actually do it for me. It
> seems
> > like stack keeps its own internal PATH, and while you can make it show it
> > to you, it prints out a bunch of other stuff that you'd have to do some
> > shell script monkeying to do anything useful with.
> >
> > The bottom line is that without being able to call ghc from the command
> > line, a lot of these tools don't work. You'd think you could just create
> an
> > alias stack ghc -> ghc, but it seems that any command line arguments you
> > pass to stack ghc go to stack and not ghc. You could manually put stack's
> > ghc on the path, but the whole gimmick here is that it can swap this out
> at
> > will depending on the project. I'd hate to have to detect this and fix it
> > constantly.
> >
> > I feel that they'll probably figure this stuff out, but for now I'm going
> > to go back to banging my head against cabal.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/ucsd-progsys/liquidhaskell/issues/407#issuecomment-115401964
> >
>
> > .
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ucsd-progsys/liquidhaskell/issues/407#issuecomment-115448006>
> .
>



-- 
Ranjit.

**Comment by spinda** [comment]
@christetreault
 
I see. My main use case for `stack` is avoiding sandbox setup, rebuilds, keeping installs in sync, etc. when working across multiple packages, so for me personally this subset of `stack`'s features is a win.

Would you mind forwarding this in some form to the `stack` [mailing list](groups.google.com/d/forum/haskell-stack)? I'm sure they'd appreciate the feedback.

**Comment by christetreault** [comment]
Poking around today, I found this issue opened against ghc-mod:

https://github.com/kazu-yamamoto/ghc-mod/issues/498

...which links to this issue against stack:

https://github.com/commercialhaskell/stack/issues/273

Doesn't seem that anybody is interested in supporting stack/ghc-mod interop at the moment. :(

**Comment by ranjitjhala** [comment]
The bigger problem, I fear, is that `ghc-mod` is perhaps quite
complex (at least, it does a lot of things). This is partly why
I prefer `hdevtools` which gives me the key things: fast checking
and type information and,

(1) is compatible with GHC 7.10

(2) is also a lot *simpler* in that it does less, and is
    easily hackable by others, and hence, more rapidly fixable...

(But I am puzzled because I think *many* more tools use ghc-mod so
I'd think it would be more up-to-date...)


On Fri, Jun 26, 2015 at 11:02 AM, Chris Tetreault <notifications@github.com>
wrote:

> Poking around today, I found this issue opened against ghc-mod:
>
> kazu-yamamoto/ghc-mod#498
> <https://github.com/kazu-yamamoto/ghc-mod/issues/498>
>
> ...which links to this issue against stack:
>
> commercialhaskell/stack#273
> <https://github.com/commercialhaskell/stack/issues/273>
>
> Doesn't seem that anybody is interested in supporting stack/ghc-mod
> interop at the moment. :(
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ucsd-progsys/liquidhaskell/issues/407#issuecomment-115818528>
> .
>



-- 
Ranjit.

**Comment by christetreault** [comment]
According to the repo, ghc 7.10 support is fixed in master. Would be nice if they'd push it to github.

Honestly, I don't use most of the fancy features of ghc-mod. I just want the flycheck stuff to work. Maybe it's time I switched over...

**Comment by ranjitjhala** [comment]
I checked out the master and successfully built it with ghc710 but couldn't use it on my projects...

**Comment by christetreault** [comment]
I found this post on reddit today: https://www.reddit.com/r/haskell/comments/3cf5yd/stack_ghcmod_work_in_progress/

It looks like _somebody_ is working on stack/ghc-mod interop. Hopefully something comes of it.

**Comment by ranjitjhala** [comment]
Yes, but the odd error messages (also listed in that thread) are exactly
the problem I was running into. Still, pretty sure they'll be fixed soon!

On Tue, Jul 7, 2015 at 10:54 AM, Chris Tetreault <notifications@github.com>
wrote:

> I found this post on reddit today:
> https://www.reddit.com/r/haskell/comments/3cf5yd/stack_ghcmod_work_in_progress/
>
> It looks like *somebody* is working on stack/ghc-mod interop. Hopefully
> something comes of it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ucsd-progsys/liquidhaskell/issues/407#issuecomment-119283208>
> .
>



-- 
Ranjit.

**Comment by ranjitjhala** [comment]
This looks like its closed. Indeed, we're _using_ stack to run LH on itself...

**Comment by JeffreyBenjaminBrown** [comment]
Good to hear that it already works! So what's the best way to add Liquid Haskell to a Stack environment?

**Comment by nikivazou** [comment]
You need to add the following dependencies in your `stack.yaml` and  then run `stack install liquidhaskell`.

**Comment by JeffreyBenjaminBrown** [comment]
Which dependencies?

**Comment by ranjitjhala** [comment]
Following @Gabriel439 [blog post](http://www.haskellforall.com/2015/12/compile-time-memory-safety-using-liquid.html) I think it should be something like:

```
resolver: lts-8.9

packages: []

extra-deps:
- aeson-0.11.3.0
- csv-table-0.1.0.1
- dotgen-0.4.2
- fgl-visualize-0.1.0.1
- intern-0.9.1.4
- located-base-0.1.1.0
- liquid-fixpoint-0.7.0.2
```

and then 

```
$ stack install liquidhaskell
```

?

@nikivazou  is that right? If so we should add these instructions to the `README`.

Also relevant is @spinda 's 

https://github.com/spinda/liquidhaskell-cabal

which is awesome, but needs a little updating...

**Comment by Gabriel439** [comment]
In the long run you should consider adding `liquidhaskell` to Stackage, because then you'll get automatically notified any time you don't build against the latest version of one of your dependencies
